### PR TITLE
mgr/cephadm: Manage /etc/ceph/ceph.conf

### PR DESCRIPTION
**Pull request by sebastian-philipp**
**TODO**

- [x] replace ceph.conf handling in qa/tasks/cephadm.py
  
 https://github.com/ceph/ceph/blob/b2de27be3f6f2eaba22df31e9e45d5633c941596/qa/tasks/cephadm.py#L961-L964

- [x] pytest

**Considerations**

* This is now embedded using the usual check_* methods, which means it's using a different code path than the daemon's reconfig calls. otoh, we're not dealing with daemons here.
* I don't like `error_ok=True`, as id detours the normal python error handling. Making it necessary to add special code for it.

**To enable the management of ceph.conf**

`ceph config set mgr mgr/cephadm/manage_etc_ceph_ceph_conf true`

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/inventory.py [hunk 1]
```diff
@@ -175,6 +175,7 @@ def __init__(self, mgr):
         self.daemon_config_deps = {}   # type: Dict[str, Dict[str, Dict[str,Any]]]
         self.last_host_check = {}      # type: Dict[str, datetime.datetime]
         self.loading_osdspec_preview = set()  # type: Set[str]
+        self.etc_ceph_ceph_conf_refresh_queue: List[str] = []
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/inventory.py:178]: why do you change the kind of annotation?

File: src/pybind/mgr/cephadm/inventory.py [hunk 2]
```diff
@@ -420,6 +426,14 @@ def host_needs_check(self, host):
             seconds=self.mgr.host_check_interval)
         return host not in self.last_host_check or self.last_host_check[host] < cutoff
 
+    def host_needs_new_etc_ceph_ceph_conf(self, host):
+        if not self.mgr.manage_etc_ceph_ceph_conf:
+            return False
+        if host in self.osdspec_previews_refresh_queue:
```
Comment (jschmid1) [inline src/pybind/mgr/cephadm/inventory.py:432]: How does that correlate?

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -536,7 +544,8 @@ def config_notify(self):
         self.event.set()
 
     def notify(self, notify_type, notify_id):
-        pass
+        if notify_type == "mon_map":
```
Comment (jschmid1) [inline src/pybind/mgr/cephadm/module.py:547]: Is there any kind of documentation which type of `notify_type`s there  are?

File: src/pybind/mgr/cephadm/module.py [hunk 4]
```diff
@@ -1389,10 +1432,12 @@ def _daemon_action(self, daemon_type, daemon_id, host, action):
...
-            out, err, code = self._run_cephadm(
-                host, name, 'unit',
-                ['--name', name, a],
-                error_ok=True)
+            try:
+                out, err, code = self._run_cephadm(
+                    host, name, 'unit',
+                    ['--name', name, a])
+            except Exception:
+                self.log.exception('cephadm failed')
```
Comment (jschmid1) [inline src/pybind/mgr/cephadm/module.py:1440]: being a bit more descriptive would certainly help :)

File: src/pybind/mgr/cephadm/module.py [hunk 5]
```diff
@@ -1140,6 +1160,11 @@ def _refresh_hosts_and_daemons(self):
                 if r:
                     failures.append(r)
 
+            if self.cache.host_needs_new_etc_ceph_ceph_conf(host):
+                self.log.debug(f"deploying new /etc/ceph/ceph.conf {host}")
```
Comment (jschmid1) [inline src/pybind/mgr/cephadm/module.py:1164]:
```suggestion
                self.log.debug(f"deploying new /etc/ceph/ceph.conf to <{host}>")
```

File: src/pybind/mgr/cephadm/module.py [hunk 6]
```diff
@@ -886,6 +895,46 @@ def _executable_path(self, conn, executable):
             executable_path))
         return executable_path
 
+    @contextmanager
```
Comment (jschmid1) [inline src/pybind/mgr/cephadm/module.py:898]: +1 for using a contextmanager here

File: src/pybind/mgr/cephadm/module.py (hunk 4 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:1440]: `log.exception` will print:

- the full exceptiont traceback
- the error message of the exception

MAybe I could add the name or something.

File: src/pybind/mgr/cephadm/inventory.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/inventory.py:432]: whoops!

File: src/pybind/mgr/cephadm/inventory.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/inventory.py:178]: cause I didn't want to change the other unrelated lines.

File: src/pybind/mgr/cephadm/module.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:547]: https://docs.ceph.com/docs/master/mgr/modules/#receiving-notifications

File: src/pybind/mgr/cephadm/module.py [hunk 7]
```diff
@@ -1252,6 +1277,24 @@ def _refresh_host_devices(self, host):
         self.cache.save_host(host)
         return None
 
+    def _deploy_etc_ceph_ceph_conf(self, host: str) -> Optional[str]:
+        try:
+            with self._remote_connection(host) as tpl:
+                conn, connr = tpl
+                out, err, code = remoto.process.check(
+                    conn,
+                    ['mkdir', '-p', '/etc/ceph'])
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/module.py:1286]: what if a different output dir was used during bootstrap (e.g. vstart)?
```
$ cephadm bootstrap ... --output-dir ~/ceph/build/
```

File: src/pybind/mgr/cephadm/inventory.py [hunk 8]
```diff
@@ -420,6 +426,16 @@ def host_needs_check(self, host):
             seconds=self.mgr.host_check_interval)
         return host not in self.last_host_check or self.last_host_check[host] < cutoff
 
+    def host_needs_new_etc_ceph_ceph_conf(self, host):
+        if not self.mgr.manage_etc_ceph_ceph_conf:
+            return False
+        if self.mgr.paused:
+            return False
+        if host in self.etc_ceph_ceph_conf_refresh_queue:
+            self.etc_ceph_ceph_conf_refresh_queue.remove(host)
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/inventory.py:435]: remove from the list while preforming the action (not during the check)?

File: src/pybind/mgr/cephadm/module.py [hunk 9]
```diff
@@ -1252,6 +1277,24 @@ def _refresh_host_devices(self, host):
...
+                if code:
+                    return f'failed to create /etc/ceph on {host}: {err}'
+                out, err, code = remoto.process.check(
+                    conn,
+                    ['dd', 'of=/etc/ceph/ceph.conf'])
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/module.py:1291]: maybe I don't see it, but where are we getting the file content? stdin?

**Comment by sebastian-philipp** [comment]
> Couple thoughts/questions:
> 
> 1. seems like this would break any host with a multi-cluster setup?

definitely.

> 2. what about `ceph.client.admin.keyring` or `ceph.pub` ?

deploying the admin keyring? hm. maybe. might be an idea, but on which hosts? Placement spec? @ricardoasmarques + @fmount wdyt?

> 5. add an action specific to sync'ing `ceph.conf` on demand?

I think we need this in any case. Maybe `ceph cephadm deploy-etc-ceph-conf [<host>]`?

> 3. if the action fails should it be retried on the next host check?

I#d make this a manual step. I really don#t want to head into the realm of automatically deploying non-containerized stuff. 

> 4. what if the `ceph.conf` file was changed manually? detect this?

I'd simply overwrite it.

File: src/pybind/mgr/cephadm/inventory.py (hunk 1 above)
Comment (jmolmo) [reply src/pybind/mgr/cephadm/inventory.py]: So ... should we avoid to use this kind of annotation?
`self.etc_ceph_ceph_conf_refresh_queue = []  # type: List[str]`

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/inventory.py]: this syntax:
```
var: int = 0
```
was added recently in Python 3.6. So, yes. for the mgr module, I'd avoid the comment-style annotations.

File: src/pybind/mgr/cephadm/module.py [hunk 10]
```diff
@@ -287,6 +294,7 @@ def __init__(self, *args, **kwargs):
             self.prometheus_alerts_path = ''
             self.migration_current = None
             self.config_dashboard = True
+            self.manage_etc_ceph_ceph_conf = True
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:297]: that kills the multi-cluster support. don't know if we want that.

File: src/pybind/mgr/cephadm/module.py (hunk 7 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:1286]: Under no circumstance, I'd automatically enable `manage_etc_ceph_ceph_conf` if `output-dir` is something different than `/etc/ceph`

File: src/pybind/mgr/cephadm/module.py (hunk 9 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:1291]: that's why I said in the standup, this PR as of right now, doesn't do anything. Note the PR is still [WIP]

**Comment by fmount** [comment]
> > Couple thoughts/questions:
> > 
> > 1. seems like this would break any host with a multi-cluster setup?
> 
> definitely.
> 
> > 1. what about `ceph.client.admin.keyring` or `ceph.pub` ?
> 
> deploying the admin keyring? hm. maybe. might be an idea, but on which hosts? Placement spec? @ricardoasmarques + @fmount wdyt?

Speaking about the OpenStack context with Director deployed Ceph cluster, the mons/mgrs are 3 and they are collocated into the Controller nodes.
I should login to one of the three controllers and exec a Ceph client to run commands against the Ceph cluster, but if the Controller[0] (which is the bootstrap node in cephadm terminology) goes down, I should be able to login on the second Controller, get a ceph client (like [1]) and work with my cluster.
The ceph-ansible deployment gives this kind of experience, and I think having the keyring and ceph.conf is the minimal amount of actions that should be taken by cephadm when a new cluster is deployed or new monitors are added/scheduled to new hosts via spec.

> 
> > 1. add an action specific to sync'ing `ceph.conf` on demand?
> 
> I think we need this in any case. Maybe `ceph cephadm deploy-etc-ceph-conf [<host>]`?

Right now we have this role [2] that is able to get the relevant data from the first controller (or mon) and sync them to the other existing monitors, but we still have the problem of updating `ceph.conf` with the list of monitors, so I'm not sure it's just a matter of copying keyring and conf but also be consistent (in terms of ceph.conf) with the state of the cluster (e.g., the `mon_host` param should be updated accordingly).
This role is useful also for  clusters that are external to OpenStack and we need a client to be able to access the Ceph Cluster.

> 
> > 1. if the action fails should it be retried on the next host check?
> 
> I#d make this a manual step. I really don#t want to head into the realm of automatically deploying non-containerized stuff.

Agree with that

> 
> > 1. what if the `ceph.conf` file was changed manually? detect this?
> 
> I'd simply overwrite it.

Manual operations should be avoided and we need a way to make the config consistent, but there are useful options [3] that can be configured on a new deployment (e.g. HCI environments).
Is there a plan to have a way to customize ceph.conf somehow?



[1]  https://github.com/fmount/tripleo-ceph/blob/master/roles/tripleo_cluster_set_container_cli/tasks/set_container_cli.yaml#L13
[2] https://github.com/fmount/tripleo-ceph/blob/master/roles/ceph_client/tasks/sync.yml
[3] https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-config/templates/ceph.conf.j2

**Comment by sebastian-philipp** [comment]
> > > Couple thoughts/questions:
> > > 
> > > 1. what about `ceph.client.admin.keyring` or `ceph.pub` ?
> > 
> > 
> > deploying the admin keyring? hm. maybe. might be an idea, but on which hosts? Placement spec? @ricardoasmarques + @fmount wdyt?
> 
> Speaking about the OpenStack context with Director deployed Ceph cluster, the mons/mgrs are 3 and they are collocated into the Controller nodes.
> I should login to one of the three controllers and exec a Ceph client to run commands against the Ceph cluster, but if the Controller[0] (which is the bootstrap node in cephadm terminology) goes down, I should be able to login on the second Controller, get a ceph client (like [1]) and work with my cluster.
> The ceph-ansible deployment gives this kind of experience, and I think having the keyring and ceph.conf is the minimal amount of actions that should be taken by cephadm when a new cluster is deployed or new monitors are added/scheduled to new hosts via spec.

Interesting. Onto which hosts should the admin keyring distributed? At least I'd think this might be independent of the MONs. 

> 
> > > 1. add an action specific to sync'ing `ceph.conf` on demand?
> > 
> > 
> > I think we need this in any case. Maybe `ceph cephadm deploy-etc-ceph-conf [<host>]`?
> 
> Right now we have this role [2] that is able to get the relevant data from the first controller (or mon) and sync them to the other existing monitors, but we still have the problem of updating `ceph.conf` with the list of monitors, so I'm not sure it's just a matter of copying keyring and conf but also be consistent (in terms of ceph.conf) with the state of the cluster (e.g., the `mon_host` param should be updated accordingly).

Actually you can't properly manage the ceph.conf at all, simply as you're not getting notified when a new MON enters the cluster.

Which files are you interested in, other than the ceph.conf?

> This role is useful also for clusters that are external to OpenStack and we need a client to be able to access the Ceph Cluster.

OK, things get interesing now. If you add a new host to the cluster, cephamd will:

* deploy node-exporter
* deploy ceph-crash
* and now, it will also deploy the ceph.conf. 

Is this sufficient and ok for a client machine? Otherwise you'll need to distribute the ceph.conf yourself.

>  
> > > 1. what if the `ceph.conf` file was changed manually? detect this?
> > 
> > 
> > I'd simply overwrite it.
> 
> Manual operations should be avoided and we need a way to make the config consistent, but there are useful options [3] that can be configured on a new deployment (e.g. HCI environments).
> Is there a plan to have a way to customize ceph.conf somehow?

By this point, ceph.conf should be nearly empty and should contain only the information to access the other MONs. If you need anything, please use `ceph config` for it. 

> 
> [1] https://github.com/fmount/tripleo-ceph/blob/master/roles/tripleo_cluster_set_container_cli/tasks/set_container_cli.yaml#L13
> [2] https://github.com/fmount/tripleo-ceph/blob/master/roles/ceph_client/tasks/sync.yml
> [3] https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-config/templates/ceph.conf.j2

File: src/pybind/mgr/cephadm/module.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:1291]:
```
ll /etc/ceph/ceph.conf 
-rw-r--r-- 1 root root 181 Jun 17 14:29 /etc/ceph/ceph.conf
´´´
I think this is ok

**Comment by sebastian-philipp** [comment]
@jdurgin do you know, if we will need support for named ceph clusters in `qa/tasks/cephadm.py`? Maybe for rbd-mirror or RGW multizone? 

something like 

https://github.com/ceph/ceph/blob/b2de27be3f6f2eaba22df31e9e45d5633c941596/qa/tasks/cephadm.py#L961-L964

I'd like to avoid adding this to cephadm itself.

**Comment by sebastian-philipp** [comment]
right now, this is IMO safe to merge, as this is disabled by default using a feature-flag. enables us to improve it later on (e.g. enable it by default, change the behaviour etc)

**Comment by fmount** [comment]
> > > > Couple thoughts/questions:
> > > > 
> > > > 1. what about `ceph.client.admin.keyring` or `ceph.pub` ?
> > > 
> > > 
> > > deploying the admin keyring? hm. maybe. might be an idea, but on which hosts? Placement spec? @ricardoasmarques + @fmount wdyt?
> > 
> > 
> > Speaking about the OpenStack context with Director deployed Ceph cluster, the mons/mgrs are 3 and they are collocated into the Controller nodes.
> > I should login to one of the three controllers and exec a Ceph client to run commands against the Ceph cluster, but if the Controller[0] (which is the bootstrap node in cephadm terminology) goes down, I should be able to login on the second Controller, get a ceph client (like [1]) and work with my cluster.
> > The ceph-ansible deployment gives this kind of experience, and I think having the keyring and ceph.conf is the minimal amount of actions that should be taken by cephadm when a new cluster is deployed or new monitors are added/scheduled to new hosts via spec.
> 
> Interesting. Onto which hosts should the admin keyring distributed? At least I'd think this might be independent of the MONs.

In general OpenStack collocates monitors and mgrs in controller nodes where the ctlplane is found, so at least monitors should have all the relevant data to help operators interact w/ Ceph cluster. The undercloud isn't able to reach the StorageNetwork on the overcloud, so monitors are the first entrypoint for Director deployed clusters.
However, any node that is able to reach the Ceph cluster public network (which is set by [1]) can be configured as a client, so yes, this not depends-on the MONs and that's the reason we have a sync role [2] to copy both conf and keyring.
> 
> > > > 1. add an action specific to sync'ing `ceph.conf` on demand?
> > > 
> > > 
> > > I think we need this in any case. Maybe `ceph cephadm deploy-etc-ceph-conf [<host>]`?
> > 
> > 
> > Right now we have this role [2] that is able to get the relevant data from the first controller (or mon) and sync them to the other existing monitors, but we still have the problem of updating `ceph.conf` with the list of monitors, so I'm not sure it's just a matter of copying keyring and conf but also be consistent (in terms of ceph.conf) with the state of the cluster (e.g., the `mon_host` param should be updated accordingly).
> 
> Actually you can't properly manage the ceph.conf at all, simply as you're not getting notified when a new MON enters the cluster.
> 
> Which files are you interested in, other than the ceph.conf?
> 

Not sure right now, ceph.conf and keys are required when a new mon is scheduled, but I need to understand which is the gap compared to `ceph-ansible` to answer your question.
For now I think generate a proper ceph.conf and sync it across the mons is the minimal amount of actions that we need to make the clients working.
> > This role is useful also for clusters that are external to OpenStack and we need a client to be able to access the Ceph Cluster.
> 
> OK, things get interesing now. If you add a new host to the cluster, cephamd will:
> 
>     * deploy node-exporter
> 
>     * deploy ceph-crash
> 
>     * and now, it will also deploy the ceph.conf.
> 
> 
> Is this sufficient and ok for a client machine? Otherwise you'll need to distribute the ceph.conf yourself.

For a client node != mons it's ok, I'd like to copy and sync stuff myself since it's an extra logic for cephadm that cannot cover all the potential scenarios, what I'm not sure right now it's the reason why you deploy node-exporter and ceph-crash by default in a client node (it can be external to the Ceph cluster)
> 
> > > > 1. what if the `ceph.conf` file was changed manually? detect this?
> > > 
> > > 
> > > I'd simply overwrite it.
> > 
> > 
> > Manual operations should be avoided and we need a way to make the config consistent, but there are useful options [3] that can be configured on a new deployment (e.g. HCI environments).
> > Is there a plan to have a way to customize ceph.conf somehow?
> 
> By this point, ceph.conf should be nearly empty and should contain only the information to access the other MONs. If you need anything, please use `ceph config` for it.

Ack, I'll try to push config using this approach
> 
> > [1] https://github.com/fmount/tripleo-ceph/blob/master/roles/tripleo_cluster_set_container_cli/tasks/set_container_cli.yaml#L13
> > [2] https://github.com/fmount/tripleo-ceph/blob/master/roles/ceph_client/tasks/sync.yml
> > [3] https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-config/templates/ceph.conf.j2

[1] https://github.com/fmount/tripleo-ceph/blob/master/roles/tripleo_cluster_mon_config/tasks/set_monitor_public_network.yaml
[2] https://github.com/fmount/tripleo-ceph/blob/master/roles/ceph_client/tasks/sync.yml

**Comment by sebastian-philipp** [comment]
green: http://pulpito.ceph.com/swagner-2020-06-23_13:15:09-rados:cephadm-wip-swagner3-testing-2020-06-23-1058-distro-basic-smithi/

failures: 

* https://tracker.ceph.com/issues/46046 (NFS)
* https://tracker.ceph.com/issues/45628 (upgrade)

**Comment by sebastian-philipp** [comment]
> In general OpenStack collocates monitors and mgrs in controller nodes where the ctlplane is found, so at least monitors should have all the relevant data to help operators interact w/ Ceph cluster. The undercloud isn't able to reach the StorageNetwork on the overcloud, so monitors are the first entrypoint for Director deployed clusters.
> However, any node that is able to reach the Ceph cluster public network (which is set by [1]) can be configured as a client, so yes, this not depends-on the MONs and that's the reason we have a sync role [2] to copy both conf and keyring.

sounds reasonable.

> 
> 
> Not sure right now, ceph.conf and keys are required when a new mon is scheduled, but I need to understand which is the gap compared to `ceph-ansible` to answer your question.
> For now I think generate a proper ceph.conf and sync it across the mons is the minimal amount of actions that we need to make the clients working.

ok. I'm going to merge this as it is.

I think we can consider adding new things to distribute later on.

> 
> > > This role is useful also for clusters that are external to OpenStack and we need a client to be able to access the Ceph Cluster.
> > 
> > 
> For a client node != mons it's ok, I'd like to copy and sync stuff myself since it's an extra logic for cephadm that cannot cover all the potential scenarios, what I'm not sure right now it's the reason why you deploy node-exporter and ceph-crash by default in a client node (it can be external to the Ceph cluster)

that can be configured by changing the placement specification of the services. Deploying them on **all** known hosts is the default right now.

**Comment by sebastian-philipp** [comment]
jenkins retest this

**Comment by mgfritch** [comment]
jenkins test make check

**Comment by mgfritch** [comment]
jenkins test docs

**Comment by mgfritch** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-06-29_09:26:42-rados:cephadm-wip-swagner-testing-2020-06-26-1524-distro-basic-smithi/

* Failure: https://tracker.ceph.com/issues/46178

**Pull request merged**